redsky/streamer.py

```diff
@@ -0,0 +1,41 @@
+from copy import deepcopy as dc
+
+
+def db_store(db, fs_data_name_save_map=None):
+    if fs_data_name_save_map is None:
+        fs_data_name_save_map = {}
+
+    def wrap(f):
+        def wrapped_f(*args, **kwargs):
+            gen = f(*args, **kwargs)
+            for name, doc in gen:
+                fs_doc = dc(doc)
+
+                if name == 'descriptor':
+                    # Mutate the doc here to handle filestore
+                    for data_name in fs_data_name_save_map.keys():
+                        fs_doc['data_keys'][data_name].update(
+                            external='FILESTORE:')
+                    fs_doc.update(
+                        filled={k: False for k in fs_data_name_save_map.keys()}
+                    )
+
+                    doc.update(
+                        filled={k: True for k in fs_data_name_save_map.keys()})
+
+                elif name == 'event':
+                    # Mutate the doc here to handle filestore
+                    for data_name, sub_dict in fs_data_name_save_map.items():
+                        fs_doc = sub_dict['saver'](data_name, fs_doc,
```

Comment:
The decorator needs to be able to ask the saver for a new resource when appropriate and then for new datum(s) for that resource. That is, the saver needs to provide two entry points (corresponding to insert_resource and insert_datum) and coordinate state between them (the current resource of interest).

Unfortunately, this is a job for a class. For the sake of argument, one could use a *pair* of simple functions (resource_saver, datum_saver) that close over the same state, but now the job of coordinating it all leaks outside the saver... better to just use a class.

Proposed API:

```python
class Saver:
    """
    Instantiating this class creates a new resource. Writing adds datums to that resource.
    """
    SPEC = '...'
    def __init__(self, fs, root):
        self._root = root
        self._closed = False
        self._fs = fs
        self._fp =  # generate filepath using self._root and random name generation
        self._current_resource = self._fs.insert_resource(self.SPEC, fp, resource_kwargs)
    def write(self, data):
        "Save data to file, generate new datum_id, insert datum, return datum_id."
        if self._closed:
            raise RuntimeError('new_resource must be called first')
        # save the data to self._fp
        datum_id = str(uuid.uuid4())
        self._fs.insert_datum(resource=self._current_resource, datum_id=datum_id,
                              datum_kwargs=datum_kwargs)
        return datum_id
    def close(self):
        # make it impossible to accidentally write more datums to this resource
        self._closed = True
        # close file handle, if applicable
    def __enter__(self):
        return self
    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()
```

Usage:
```python
s = Saver(db.fs, '/tmp/mydirectory')
s.write(array1)
s.write(array2)
s.close()

# or, equivalently:

with Saver(db.fs, '/tmp/mydirectory') as s:
    s.write(array1)
    s.write(array2)
```

We'll need a Saver class per each SPEC we want to write. A given Saver instance binds to exactly one resource during its lifetime. Once it is closed it cannot be reused. (To get a new resource, instantiate another `Saver`....)

I just ran this by @tacaswell and got a :+1:. We really wanted this to be possible with simple functions, but it looks like this is as simple as things can get.